src/python/pants/core/goals/update_build_files.py

```diff
@@ -380,11 +380,118 @@ def should_be_renamed(token: tokenize.TokenInfo) -> bool:
     )
 
 
+# ------------------------------------------------------------------------------------------
+# Rename deprecated field types fixer
+# ------------------------------------------------------------------------------------------
+
+
+class RenameDeprecatedFieldsRequest(DeprecationFixerRequest):
+    pass
+
+
+class RenamedFieldTypes(FrozenDict[Tuple[str, str], str]):
+    """Deprecated field type names for target to new names."""
```

Comment:
How about `FrozenDict[str, FrozenDict[str, str]]`, which maps target types to field renames. Then you can rewrite the implementation I think to more eagerly exit if the target type is not a match? 

I'm not sure I followed the level code correctly...the level should be 1. A nested target should not exist - targets are always top-level.

At this point, might also be worth switching to a frozen dataclass rather than subclassing FrozenDict so that you can give a name to the variable like `target_types_to_field_renames`.

Reply:
> How about FrozenDict[str, FrozenDict[str, str]], which maps target types to field renames. Then you can rewrite the implementation I think to more eagerly exit if the target type is not a match?

I think you're referring to eagerly exit out of `should_be_renamed`, correct?
OK, will try it out.. see what it'll look like (not convinced it will win much, but perhaps the data structure will be easier to comprehend?)

Reply:
> I'm not sure I followed the level code correctly...the level should be 1. A nested target should not exist - targets are always top-level.

Agreed, there's a check that we only ever operate when `level==1`.

This is to protect against constructs, such as this, as an example:
```
python_distribution(
  provides=setup_py(my_field="value"),
)
```

As we do not want to change `my_field` here... The level code might be overkill, but it plugs a potential hole.
And, as I write this, I realize I don't only look at the field name, but the enclosing target name as well (`setup_py` in this example) and as such, it would be safe to ignore the level all together.. thanks for the nudge, will get rid of it.

Reply:
Oh, wait, no, the level is important for this reason..
```
python_distribution(
  name="dist",
  provides=setup_py(....),
  old_field="value",
)
```

We must keep track of when we're seeing fields in `setup_py`, which we should ignore, and when we're back in `python_distritbution`, so we know that `old_field` is deprecated and should be replaced.

It's hard to make this distinction unless I keep track of when I enter and exit constructs, such as `setup_py` there..

Reply:
I've made some refactorings that will hopefully be easier to follow now.